File: C++/bms.cpp
#include <iostream>
#include <vector>
#include <string>
#include <numeric>
#include <algorithm>

// Function to find the index of a city
int indexOf(const std::vector<std::string>& arr, const std::string& city) {
    for (size_t i = 0; i < arr.size(); ++i) {
        if (arr[i] == city) {
            return i;
        }
    }
    return -1;
}

// Function to print a path and its distance
void printPath(const std::vector<std::string>& path, int distance) {
    for (size_t i = 0; i < path.size(); ++i) {
        std::cout << path[i];
        if (i < path.size() - 1) {
            std::cout << "->";
        }
    }
    std::cout << "| distance = " << distance << std::endl;
}

// Function to calculate the total distance of a path
int calcDistance(const std::vector<std::string>& path, const std::vector<std::string>& cities, const std::vector<std::vector<int>>& graph) {
    int total = 0;
    for (size_t i = 0; i < path.size() - 1; ++i) {
        int from = indexOf(cities, path[i]);
        int to = indexOf(cities, path[i + 1]);
        if (from != -1 && to != -1) {
            total += graph[from][to];
        }
    }
    return total;
}

// Function to validate if a path is connected in the graph
bool pathValid(const std::vector<std::string>& path, const std::vector<std::string>& cities, const std::vector<std::vector<int>>& graph) {
    for (size_t i = 0; i < path.size() - 1; ++i) {
        int from = indexOf(cities, path[i]);
        int to = indexOf(cities, path[i + 1]);
        if (from == -1 || to == -1 || graph[from][to] == 0) {
            return false;
        }
    }
    return true;
}

// Recursive function to generate all permutations of a list
void perm(std::vector<std::string>& arr, int index, std::vector<std::vector<std::string>>& results) {
    if (index == arr.size()) {
        results.push_back(arr);
        return;
    }
    for (size_t i = index; i < arr.size(); ++i) {
        std::swap(arr[i], arr[index]);
        perm(arr, index + 1, results);
        std::swap(arr[i], arr[index]);
    }
}

// Wrapper function for permutations
std::vector<std::vector<std::string>> genPerm(const std::vector<std::string>& list) {
    std::vector<std::vector<std::string>> results;
    if (list.empty()) {
        results.push_back({});
        return results;
    }
    std::vector<std::string> arr = list;
    perm(arr, 0, results);
    return results;
}

// Recursive function to generate subsets of a given size
void genSubHelp(const std::vector<std::string>& list, int k, size_t start, std::vector<std::string>& current, std::vector<std::vector<std::string>>& result) {
    if (current.size() == k) {
        result.push_back(current);
        return;
    }
    for (size_t i = start; i < list.size(); ++i) {
        current.push_back(list[i]);
        genSubHelp(list, k, i + 1, current, result);
        current.pop_back();
    }
}

// Wrapper function for subset generation
std::vector<std::vector<std::string>> genSub(const std::vector<std::string>& list, int k) {
    std::vector<std::vector<std::string>> result;
    std::vector<std::string> current;
    genSubHelp(list, k, 0, current, result);
    return result;
}

int main() {
    std::vector<std::string> cities = {"S", "A", "B", "C", "D", "E", "G"};
    std::vector<std::vector<int>> graph = {
        {0, 1, 1, 0, 0, 0, 0},
        {1, 0, 1, 0, 1, 0, 0},
        {1, 1, 0, 1, 0, 0, 0},
        {0, 0, 1, 0, 0, 1, 0},
        {0, 1, 0, 0, 0, 0, 1},
        {0, 0, 0, 1, 0, 0, 0},
        {0, 0, 0, 0, 1, 0, 0}
    };

    std::string startcit = "S";
    std::string endcit = "G";

    std::vector<std::string> middle;
    for (const std::string& city : cities) {
        if (city != startcit && city != endcit) {
            middle.push_back(city);
        }
    }
    
    std::cout << "All valid paths from " << startcit << " to " << endcit << std::endl;
    
    for (size_t size = 0; size <= middle.size(); ++size) {
        std::vector<std::vector<std::string>> subsets = genSub(middle, size);
        for (const auto& subset : subsets) {
            std::vector<std::vector<std::string>> midpaths = genPerm(subset);
            
            for (const auto& midpath : midpaths) {
                std::vector<std::string> fullpath;
                fullpath.push_back(startcit);
                fullpath.insert(fullpath.end(), midpath.begin(), midpath.end());
                fullpath.push_back(endcit);
                
                if (pathValid(fullpath, cities, graph)) {
                    int distance = calcDistance(fullpath, cities, graph);
                    printPath(fullpath, distance);
                }
            }
        }
    }

    return 0;
}

File: Java/bms.java
import java.util.*;

public class BMS{
    public static void main(String[] args){
        String[] cities = {"S","A","B","C","D","E","G"};

        int[][] graph = {
            {0,1,1,0,0,0,0},
            {1,0,1,0,1,0,0},
            {1,1,0,1,0,0,0},
            {0,0,1,0,0,1,0},
            {0,1,0,0,0,0,1},
            {0,0,0,1,0,0,0},
            {0,0,0,0,1,0,0}
        };

        String startcit = "S";
        String endcit = "G";

        List<String> middle = new ArrayList<>();
        for(String city : cities){
            if(!city.equals(startcit) && !city.equals(endcit)){
                middle.add(city);
            }
        }
        
        System.out.println("All valid paths from "+startcit+" to "+endcit);
        
        for(int size =0;size<=middle.size();size++){
            List<List<String>> subsets = gensub(middle,size);
            for(List<String> subset : subsets){
                List<List<String>> midpaths = Genperm(subset);
        
                for(List<String>midpath:midpaths)
                {
                    List<String> fullpath = new ArrayList<>();
                    fullpath.add(startcit);
                    fullpath.addAll(midpath);
                    fullpath.add(endcit);
        
                    if(Pathvalid(fullpath,cities,graph)){
                        int distance = calcdistance(fullpath,cities,graph);
                        printpath(fullpath,distance);
                    }
                }
            }
        }
    }
    private static List<List<String>> gensub(List<String> list,int k){
        List<List<String>> result = new ArrayList<>();
        gensubhelp(list,k,0,new ArrayList<>(),result);
        return result;
    }
    private static void gensubhelp(List<String> list,int k,int start,List<String> current, List<List<String>> result){
        if(current.size()==k){
            result.add(new ArrayList<>(current));
            return;
        }
        for (int i = start; i < list.size(); i++) {
            current.add(list.get(i));
            gensubhelp(list, k, i + 1, current, result);
            current.remove(current.size() - 1);
        }
    }
    private static List<List<String>> Genperm(List<String> list){
        List<List<String>> results=new ArrayList<>();
        perm(list,0,results);
        return results;
    }
    private static void perm(List<String> arr,int index,List<List<String>> results)
    {
        if(index == arr.size()-1){
            results.add(new ArrayList<>(arr));
            return;
        }
        for(int i=index;i<arr.size();i++){
            Collections.swap(arr,i,index);
            perm(arr,index+1,results);
            Collections.swap(arr,i,index);
        }
    }
    private static boolean Pathvalid(List<String> path,String[] cities,int[][] graph){
        for(int i=0;i<path.size()-1;i++){
            int from = indexOf(cities,path.get(i));
            int to = indexOf(cities,path.get(i+1));
            if(graph[from][to]==0)return false;
        }
        return true;
    }
    private static int calcdistance(List<String> path,String[] cities,int[][] graph){
        int total = 0;
        for(int i=0;i<path.size()-1;i++){
            int from = indexOf(cities,path.get(i));
            int to = indexOf(cities,path.get(i+1));
            total += graph[from][to]; 
        }
        return total;
    }
    private static int indexOf(String[] arr,String city){
        for(int i=0;i<arr.length;i++){
            if(arr[i].equals(city)) return i;
        }
        return -1;
    }

    private static void printpath(List<String> path,int distance){
        System.out.println(String.join("->",path)+"| distance = "+distance);
    }
}